### DISCUSSION TITLE
Losing HTML classes in HTML to HTML conversion

### DISCUSSION THREAD

**Discussion by dubsarGimillu** [jgm/pandoc#9318] [Q&A]
My understanding is that the pandoc-discuss mailing list is frozen after a spam attack in December and that this is the preferred place for questions.  If that is incorrect please let me know.  

I run a static site generated from HTML fragments and Pandoc templates.  I could keep HTML classes on `table` and `td` elements when I generated a page in 2022, but when I rebuild the same page today those attributes of the input HTML get erased in the output HTML.  Is there a guide to which parts of a HTML file are lost when passing through Pandoc and its abstract model of documents?  Has the HTML table > Pandoc > HTML table conversion changed since summer 2022?

Eg. today input 

```
<table class="aClass">
  <thead>
  <tr>
    <th>Column 1</th>
    <th>Column 2</th>
  </tr>
  </thead>
  <tbody>
  <tr>
    <td class="anotherClass">a</td>
    <td>b</td>
  </tr>
  <tr>
    <td class="anotherClass">x</td>
    <td>y</td>
  </tr>
  </tbody>
</table> 
```

becomes output:
```

<table>
<thead>
<tr class="header">
<th>Column 1</th>
<th>Column 2</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>b</td>
<td>b</td>
</tr>
<tr class="even">
<td>x</td>
<td>y</td>
</tr>
</tbody>
</table>
```

In 2022 the "aClass" and "anotherClass" were preserved in the output.

**Comment by jgm** [comment]
Yes, there have been relevant changes since summer 2022 (see changelog for details -- too many to list or recall).
It's possible that previously the table was just being parsed as a raw HTML block and transmitted verbatim.

We can look into preserving these attributes; why don't you submit an issue.

But generally pandoc will be lossy for HTML -> HTML conversions.  If you're going from HTML -> HML, better to skip the conversion step?

**Reply by dubsarGimillu** [reply]
Thank you!  I tried adding `--from=html+raw_html` to the pandoc command but that does not stop the lossy conversion (eg. alternate rows of tables are still assigned `odd` and even classes).  `raw_attribute` seems for marked-up fragments of HTML within a larger file not a whole file of nothing but HTML.  Is there another extension which I am missing?

**Reply by jgm** [reply]
You are correct. But as pandoc's ability to process HTML tables has increased, fewer tables are treated as undigestible and therefore raw.

**Comment by GunaPalanivel** [comment]
dubsarGimillu, a user of the Pandoc document converter, is experiencing data loss when converting HTML fragments containing tables to HTML. Specifically, classes assigned to table and td elements are being removed in the output.

Cause of the Problem:

Changes in Pandoc since summer 2022 have improved its ability to process HTML tables. This results in tables being interpreted and potentially losing attributes like classes.
Pandoc generally performs lossy conversions for HTML -> HTML, meaning some information might be lost.

Solutions Discussed:

Using --from=html+raw_html flag: While this flag preserves raw HTML blocks, it didn't work in this case as alternate row classes (odd and even) were still assigned.
Using the raw_attribute extension: This extension only applies to marked-up fragments within a larger document, not an entire HTML file.

Possible Next Steps:

Open an issue on Pandoc: dubsarGimillu is encouraged to submit an issue on the Pandoc Github repository to request preserving user-defined classes on tables.

Alternative solution: 

Consider skipping the Pandoc conversion for HTML -> HTML if preserving the specific classes is crucial.
Explore other extensions: Investigate if any other extensions exist that specifically address preserving HTML attributes during Pandoc conversions.

Additional Information:

The provided changelog for relevant changes since summer 2022 might offer more details about the table parsing improvements.
The Pandoc documentation can provide further information on flags and extensions.
